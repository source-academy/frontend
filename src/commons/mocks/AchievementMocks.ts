import {
  AchievementGoal,
  AchievementItem,
  GoalType
} from '../../features/achievement/AchievementTypes';
import { AND, EventTypes } from '../../features/achievement/ExpressionTypes';

export const mockAchievements: AchievementItem[] = [
  {
    uuid: '0',
    title: 'Rune Master',
    xp: 100,
    isVariableXp: false,
    isTask: true,
    position: 1,
    prerequisiteUuids: ['2', '1'],
    goalUuids: ['0'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/rune-master.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Rune Master achievement'
    }
  },
  {
    uuid: '1',
    title: 'Beyond the Second Dimension',
    xp: 100,
    isVariableXp: false,
    deadline: new Date(2020, 7, 6, 12, 30, 0),
    isTask: false,
    position: 0,
    prerequisiteUuids: [],
    goalUuids: ['1'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/btsd.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Beyond the Second Dimension achievement'
    }
  },
  {
    uuid: '2',
    title: 'Colorful Carpet',
    xp: 100,
    isVariableXp: false,
    deadline: new Date(2020, 7, 8, 9, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteUuids: [],
    goalUuids: ['2'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/colorful-carpet.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Colorful Carpet achievement'
    }
  },
  {
    uuid: '3',
    title: '',
    xp: 100,
    isVariableXp: false,
    isTask: false,
    position: 0,
    prerequisiteUuids: [],
    goalUuids: [],
    cardBackground: '',
    view: {
      coverImage: '',
      description: '',
      completionText: ''
    }
  },
  {
    uuid: '4',
    title: 'Curve Wizard',
    xp: 100,
    isVariableXp: false,
    deadline: new Date(2020, 8, 15, 0, 0, 0),
    isTask: true,
    position: 4,
    prerequisiteUuids: ['5', '6'],
    goalUuids: ['3'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/curve-wizard.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Wizard achievement'
    }
  },
  {
    uuid: '5',
    title: 'Curve Introduction',
    xp: 100,
    isVariableXp: false,
    deadline: new Date(2020, 7, 28, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteUuids: [],
    goalUuids: ['4'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/curve-introduction.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Introduction achievement'
    }
  },
  {
    uuid: '6',
    title: 'Curve Manipulation',
    xp: 100,
    isVariableXp: false,
    deadline: new Date(2020, 8, 5, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteUuids: [],
    goalUuids: ['5'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/curve-manipulation.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Manipulation achievement'
    }
  },
  {
    uuid: '21',
    title: 'The Source-rer',
    xp: 100,
    isVariableXp: false,
    deadline: new Date(2020, 7, 21, 0, 0, 0),
    isTask: true,
    position: 3,
    prerequisiteUuids: [],
    goalUuids: ['16', '18'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/the-source-rer.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed The Source-rer achievement'
    }
  },
  {
    uuid: '8',
    title: 'Power of Friendship',
    xp: 100,
    isVariableXp: false,
    isTask: true,
    position: 2,
    prerequisiteUuids: ['9'],
    goalUuids: [],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/power-of-friendship.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Power of Friendship achievement'
    }
  },
  {
    uuid: '9',
    title: 'Piazza Guru',
    xp: 100,
    isVariableXp: false,
    isTask: false,
    position: 0,
    prerequisiteUuids: [],
    goalUuids: ['8'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/piazza-guru.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Piazza Guru achievement'
    }
  },
  {
    uuid: '16',
    title: "That's the Spirit",
    xp: 100,
    isVariableXp: false,
    isTask: true,
    position: 5,
    prerequisiteUuids: [],
    goalUuids: ['14'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/thats-the-spirit.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: "Congratulations! You have completed That's the Spirit achievement"
    }
  },
  {
    uuid: '13',
    title: 'Kool Kidz',
    xp: 100,
    isVariableXp: false,
    isTask: true,
    position: 6,
    prerequisiteUuids: [],
    goalUuids: ['11'],
    cardBackground:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-background/default.png',
    view: {
      coverImage:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/cover-image/default.png',
      description: 'Diz for teh K00L K1dz',
      completionText: 'U SO G00D'
    }
  }
];

export const mockGoals: AchievementGoal[] = [
  {
    uuid: '0',
    text: 'Bonus for completing Colorful Carpet & Beyond the Second Dimension Achievements',
    meta: {
      type: GoalType.BINARY,
      condition: AND(
        { event: EventTypes.ASSESSMENT_GRADING, restriction: 'M2A' },
        { event: EventTypes.ASSESSMENT_GRADING, restriction: 'M2B' }
      ),
      targetCount: 100
    },
    count: 0,
    targetCount: 100,
    completed: false,
    achievementUuids: ['0']
  },
  {
    uuid: '1',
    text: 'XP earned from Beyond the Second Dimension Mission',
    meta: {
      type: GoalType.ASSESSMENT,
      assessmentNumber: 1,
      requiredCompletionFrac: 0.5
    },
    count: 213,
    targetCount: 250,
    completed: true,
    achievementUuids: ['1']
  },
  {
    uuid: '2',
    text: 'XP earned from Colorful Carpet Mission',
    meta: {
      type: GoalType.ASSESSMENT,
      assessmentNumber: 2,
      requiredCompletionFrac: 0.8
    },
    count: 0,
    targetCount: 250,
    completed: false,
    achievementUuids: ['2']
  },
  {
    uuid: '3',
    text: 'Bonus for completing Curve Introduction & Curve Manipulation Achievements',
    meta: {
      type: GoalType.BINARY,
      condition: AND(
        {
          event: EventTypes.ASSESSMENT_GRADING,
          restriction: 'M3'
        },
        {
          event: EventTypes.ASSESSMENT_GRADING,
          restriction: 'M4A'
        }
      ),
      targetCount: 100
    },
    count: 0,
    targetCount: 100,
    completed: false,
    achievementUuids: ['4']
  },
  {
    uuid: '4',
    text: 'XP earned from Curve Introduction Mission',
    meta: {
      type: GoalType.ASSESSMENT,
      assessmentNumber: 3,
      requiredCompletionFrac: 0.5
    },
    count: 178,
    targetCount: 250,
    completed: true,
    achievementUuids: ['5']
  },
  {
    uuid: '5',
    text: 'XP earned from Curve Manipulation Mission',
    meta: {
      type: GoalType.ASSESSMENT,
      assessmentNumber: 4,
      requiredCompletionFrac: 0.8
    },
    count: 191,
    targetCount: 250,
    completed: false,
    achievementUuids: ['6']
  },
  {
    uuid: '16',
    text: 'Submit Source 3 Path',
    meta: {
      type: GoalType.BINARY,
      condition: {
        event: EventTypes.ASSESSMENT_SUBMISSION,
        restriction: 'P3'
      },
      targetCount: 100
    },
    count: 100,
    targetCount: 100,
    completed: true,
    achievementUuids: ['21']
  },
  {
    uuid: '18',
    text: 'XP earned from Source 3 Path',
    meta: {
      type: GoalType.ASSESSMENT,
      assessmentNumber: 5,
      requiredCompletionFrac: 1
    },
    count: 300,
    targetCount: 300,
    completed: true,
    achievementUuids: ['21']
  },
  {
    uuid: '8',
    text: 'Each Top Voted answer in Piazza gives 10 XP',
    meta: {
      type: GoalType.MANUAL,
      targetCount: 100
    },
    count: 40,
    targetCount: 100,
    completed: false,
    achievementUuids: ['9']
  },
  {
    uuid: '14',
    text: 'Submit 1 PR to Source Academy Github',
    meta: {
      type: GoalType.MANUAL,
      targetCount: 100
    },
    count: 100,
    targetCount: 100,
    completed: true,
    achievementUuids: ['16']
  },
  {
    uuid: '11',
    text: 'Be the Koolest Kidz in SOC by redeeming this 100 XP achievement yourself',
    meta: {
      type: GoalType.BINARY,
      condition: false,
      targetCount: 100
    },
    count: 0,
    targetCount: 100,
    completed: false,
    achievementUuids: ['13']
  }
];
