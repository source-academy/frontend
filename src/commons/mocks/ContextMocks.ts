import { parse } from 'acorn';
import { FunctionExpression, Node } from 'estree';
import { ACORN_PARSE_OPTIONS } from 'js-slang/dist/constants';
import createContext, { EnvTree } from 'js-slang/dist/createContext';
import Closure from 'js-slang/dist/interpreter/closure';
import { Context, Environment } from 'js-slang/dist/types';
import { TypeError } from 'js-slang/dist/utils/rttc';

export function mockContext(chapter = 1): Context {
  return createContext(chapter);
}

export function mockRuntimeContext(): Context {
  const context = createContext();
  // Note: noticed no harm in removing the following context.runtime.
  // If you get an error with head is undefined after trying to evaluate code.
  // Likely due to the environments: [].
  // In every real context, there is at least one env frame.
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environmentTree: new EnvTree(),
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ],
    agenda: null,
    stash: null,
    envStepsTotal: 0,
    breakpointSteps: []
  };
  return context;
}

export function mockClosure(): Closure {
  return new Closure({} as FunctionExpression, {} as Environment, {} as Context);
}

export function mockTypeError(): TypeError {
  // Typecast to Node to fix estree-acorn compatability.
  return new TypeError(parse('', ACORN_PARSE_OPTIONS) as Node, '', '', '');
}
