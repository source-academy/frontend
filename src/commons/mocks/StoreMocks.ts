import _ from 'lodash';
import { DeepPartial, Store } from 'redux';
import mockStore from 'redux-mock-store';

import {
  defaultAcademy,
  defaultAchievement,
  defaultApplication,
  defaultDashboard,
  defaultFileSystem,
  defaultPlayground,
  defaultSession,
  defaultWorkspaceManager,
  OverallState
} from '../application/ApplicationTypes';

export function mockInitialStore(overrides?: DeepPartial<OverallState>): Store<OverallState> {
  const createStore = (mockStore as any)();
  const state: OverallState = {
    academy: defaultAcademy,
    achievement: defaultAchievement,
    application: defaultApplication,
    dashboard: defaultDashboard,
    playground: defaultPlayground,
    workspaces: defaultWorkspaceManager,
    session: defaultSession,
    fileSystem: defaultFileSystem
  };
  return createStore(_.merge(state, overrides));
}
