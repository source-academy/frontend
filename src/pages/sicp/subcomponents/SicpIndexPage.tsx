import { H1, H2, H4 } from '@blueprintjs/core';
import React from 'react';

import SicpToc from './SicpToc';

const originalAuthors = 'Harold Abelson and Gerald Jay Sussman';
const originalWithAuthors = 'with Julie Sussman';
const adaptedAuthors = 'Martin Henz and Tobias Wrigstad';
const adaptedWithAuthors = 'with Julie Sussman';
const developers = 'Samuel Fang';

const authors = (
  <div className="sicp-authors">
    <H4>{originalAuthors}</H4>
    <p>
      {originalWithAuthors}
      <i>— original authors</i>
    </p>
    <H4>{adaptedAuthors}</H4>
    <p>
      {adaptedWithAuthors}
      <i>— adapters to JavaScript</i>
    </p>
    <H4>{developers}</H4>
    <p>
      <i>— designer and developer of this Interactive SICP JS edition</i>
    </p>
  </div>
);

const bookTitle = (
  <div>
    <H1>Structure and Interpretation of Computer Programs</H1>
    <H2>— JavaScript Edition</H2>
  </div>
);

const licenses = (
  <div className="sicp-licenses">
    <div>
      <a href="http://creativecommons.org/licenses/by-sa/4.0/" rel="nofollow">
        <img src="https://licensebuttons.net/l/by-sa/4.0/88x31.png" alt="CC BY-SA 4.0" />
      </a>
    </div>
    <div>
      <p>
        This work is licensed under a{' '}
        <a href="http://creativecommons.org/licenses/by-sa/4.0/" rel="nofollow">
          Creative Commons Attribution-ShareAlike 4.0 International License
        </a>
        .
      </p>
    </div>
    <br />
    <div>
      <a href="https://www.gnu.org/licenses/gpl-3.0.en.html" rel="nofollow">
        <img
          src="https://camo.githubusercontent.com/46d38fe6087a9b9bdf7e45458901b818765b8391/68747470733a2f2f75706c6f61642e77696b696d656469612e6f72672f77696b6970656469612f636f6d6d6f6e732f7468756d622f372f37392f4c6963656e73655f69636f6e2d67706c2e7376672f353070782d4c6963656e73655f69636f6e2d67706c2e7376672e706e67"
          alt="GPL 3"
          data-canonical-src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/79/License_icon-gpl.svg/50px-License_icon-gpl.svg.png"
        />
      </a>
    </div>
    <div>
      <p>
        All JavaScript programs in this work are licensed under the{' '}
        <a href="https://www.gnu.org/licenses/gpl-3.0.en.html" rel="nofollow">
          GNU General Public License Version 3
        </a>
        .
      </p>
    </div>
    <br />
    <div>
      <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/" rel="nofollow">
        <img src="https://licensebuttons.net/l/by-nc-sa/4.0/88x31.png" alt="CC BY-NC-SA 4.0" />
      </a>
    </div>
    <div>
      <p>
        <a
          href="https://mitpress.mit.edu/books/structure-and-interpretation-computer-programs-1"
          rel="nofollow"
        >
          Print and Kindle versions of this work{' '}
        </a>
        are published by The MIT Press under a{' '}
        <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/" rel="nofollow">
          Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
        </a>
        .
      </p>
    </div>
    <br />
  </div>
);

const SicpIndexPage: React.FC = () => {
  return (
    <div className="sicp-index-page">
      <div className="sicp-cover">
        <img src="http://source-academy.github.io/sicp/sicpjs.png" alt="SICP" />
        <div className="sicp-cover-text">
          {bookTitle}
          {authors}
        </div>
      </div>
      <br />
      <H2>Content</H2>
      <SicpToc />
      <br />
      <H2>Licenses</H2>
      {licenses}
    </div>
  );
};

export default SicpIndexPage;
