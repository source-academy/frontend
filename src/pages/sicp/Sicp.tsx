import 'katex/dist/katex.min.css';

import { Button, Classes, NonIdealState, Spinner } from '@blueprintjs/core';
import classNames from 'classnames';
import * as React from 'react';
import { useDispatch } from 'react-redux';
import { RouteComponentProps, useHistory, useParams } from 'react-router';
import Constants from 'src/commons/utils/Constants';
import { setLocalStorage } from 'src/commons/utils/LocalStorageHelper';
import { resetWorkspace, toggleUsingSubst } from 'src/commons/workspace/WorkspaceActions';
import { parseArr, ParseJsonError } from 'src/features/sicp/parser/ParseJson';
import { getNext, getPrev } from 'src/features/sicp/TableOfContentsHelper';
import {
  readSicpSectionLocalStorage,
  setSicpSectionLocalStorage,
  SICP_CACHE_KEY,
  SICP_INDEX
} from 'src/features/sicp/utils/SicpUtils';

import SicpErrorBoundary from '../../features/sicp/errors/SicpErrorBoundary';
import getSicpError, { SicpErrorType } from '../../features/sicp/errors/SicpErrors';
import SicpIndexPage from './subcomponents/SicpIndexPage';

type SicpProps = RouteComponentProps<{}>;

const baseUrl = Constants.sicpBackendUrl + 'json/';
const extension = '.json';

// Context to determine which code snippet is active
export const CodeSnippetContext = React.createContext({
  active: '0',
  setActive: (x: string) => {}
});

const loadingComponent = <NonIdealState title="Loading Content" icon={<Spinner />} />;

const Sicp: React.FC<SicpProps> = props => {
  const [data, setData] = React.useState(<></>);
  const [loading, setLoading] = React.useState(false);
  const [active, setActive] = React.useState('0');
  const { section } = useParams<{ section: string }>();
  const parentRef = React.useRef<HTMLDivElement>(null);
  const refs = React.useRef({});
  const history = useHistory();

  const scrollRefIntoView = (ref: HTMLDivElement | null) => {
    if (!ref || !parentRef?.current) {
      return;
    }

    const parent = parentRef.current!;
    const relativeTop = window.scrollY > parent.offsetTop ? window.scrollY : parent.offsetTop;

    parent.scrollTo({
      behavior: 'smooth',
      top: ref.offsetTop - relativeTop
    });
  };

  // Handle loading of latest viewed section and fetch json data
  React.useEffect(() => {
    if (!section) {
      /**
       * Handles rerouting to the latest viewed section when clicking from
       * the main application navbar. history.replace is used to allow the
       * user to still use the browser back button to navigate the app.
       */
      history.replace(`/sicpjs/${readSicpSectionLocalStorage()}`);
      return;
    }

    if (section === SICP_INDEX) {
      setSicpSectionLocalStorage(SICP_INDEX);
      return;
    }

    setLoading(true);

    fetch(baseUrl + section + extension)
      .then(response => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(myJson => {
        try {
          const newData = parseArr(myJson, refs); // Might throw error
          setData(newData);
          setSicpSectionLocalStorage(section); // Sets local storage if valid page
        } catch (error) {
          throw new ParseJsonError(error.message);
        }
      })
      .catch(error => {
        console.error(error);

        if (error.message === 'Not Found') {
          // page not found
          setData(getSicpError(SicpErrorType.PAGE_NOT_FOUND_ERROR));
        } else if (error instanceof ParseJsonError) {
          // error occurred while parsing JSON
          setData(getSicpError(SicpErrorType.PARSING_ERROR));
        } else {
          setData(getSicpError(SicpErrorType.UNEXPECTED_ERROR));
        }
        setLocalStorage(SICP_CACHE_KEY, SICP_INDEX); // Prevents caching invalid page
      })
      .finally(() => {
        setLoading(false);
      });
  }, [section, history]);

  // Scroll to correct position
  React.useEffect(() => {
    if (loading) {
      return;
    }

    const hash = props.location.hash;
    const ref = refs.current[hash];

    scrollRefIntoView(ref);
  }, [props.location.hash, loading]);

  // Close all active code snippet when new page is loaded
  React.useEffect(() => {
    setActive('0');
  }, [data]);

  const dispatch = useDispatch();
  const handleSnippetEditorOpen = (s: string) => {
    setActive(s);
    dispatch(resetWorkspace('sicp'));
    dispatch(toggleUsingSubst(false, 'sicp'));
  };
  const handleNavigation = (sect: string) => {
    history.push('/sicpjs/' + sect);
  };

  const navigationButtons = (
    <div className="sicp-navigation-buttons">
      {getPrev(section) && (
        <Button onClick={() => handleNavigation(getPrev(section)!)}>Previous</Button>
      )}
      {getNext(section) && (
        <Button onClick={() => handleNavigation(getNext(section)!)}>Next</Button>
      )}
    </div>
  );

  return (
    <div
      className={classNames('Sicp', Classes.RUNNING_TEXT, Classes.TEXT_LARGE, Classes.DARK)}
      ref={parentRef}
    >
      <SicpErrorBoundary>
        <CodeSnippetContext.Provider value={{ active: active, setActive: handleSnippetEditorOpen }}>
          {loading ? (
            <div className="sicp-content">{loadingComponent}</div>
          ) : section === 'index' ? (
            <SicpIndexPage />
          ) : (
            <div className="sicp-content">
              {data}
              {navigationButtons}
            </div>
          )}
        </CodeSnippetContext.Provider>
      </SicpErrorBoundary>
    </div>
  );
};

export default Sicp;
