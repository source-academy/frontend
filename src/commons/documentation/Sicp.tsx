import 'katex/dist/katex.min.css';

import { Button, Classes, NonIdealState, Spinner, TreeNodeInfo } from '@blueprintjs/core';
import classNames from 'classnames';
import React, { useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import { Link, useNavigate } from 'react-router';
import Constants from 'src/commons/utils/Constants';
import { setLocalStorage } from 'src/commons/utils/LocalStorageHelper';
import WorkspaceActions from 'src/commons/workspace/WorkspaceActions';
import { parseArr, ParseJsonError } from 'src/features/sicp/parser/ParseJson';
import { getNext, getPrev } from 'src/features/sicp/TableOfContentsHelper';
import {
  setSicpSectionLocalStorage,
  SICP_CACHE_KEY,
  SICP_INDEX
} from 'src/features/sicp/utils/SicpUtils';

import SicpErrorBoundary from '../../features/sicp/errors/SicpErrorBoundary';
import getSicpError, { SicpErrorType } from '../../features/sicp/errors/SicpErrors';
import SicpIndexPage from './SicpIndexPage';
import SicpNavigationBar from './SicpNavigationBar';

const baseUrl = Constants.sicpBackendUrl + 'json/';
const extension = '.json';

// Context to determine which code snippet is active
export const CodeSnippetContext = React.createContext({
  active: '0',
  setActive: (x: string) => {}
});

const loadingComponent = <NonIdealState title="Loading Content" icon={<Spinner />} />;

type SicpProps = {
  setSicpHomeCallBackFn: (fn: () => void) => void;
};

const Sicp: React.FC<SicpProps> = props => {
  const [data, setData] = useState(<></>);
  const [loading, setLoading] = useState(false);
  const [active, setActive] = useState('0');
  const [section, setSection] = useState('index');
  const [hash, setHash] = useState('#begin');
  const refs = useRef<Record<string, HTMLElement | null>>({});
  const navigate = useNavigate();

  const scrollRefIntoView = (ref: HTMLElement | null) => {
    // console.log(`Scrolling ${ref} into view...`);
    if (!ref) {
      return;
    }

    ref.scrollIntoView({
      behavior: 'smooth'
    });
  };

  // Handle loading of latest viewed section and fetch json data
  React.useEffect(() => {
    props.setSicpHomeCallBackFn(() => {
      setSection('index');
    });
    if (!section) {
      /**
       * Handles rerouting to the latest viewed section when clicking from
       * the main application navbar. Navigate replace logic is used to allow the
       * user to still use the browser back button to navigate the app.
       */
      // navigate(`/sicpjs/${readSicpSectionLocalStorage()}`, { replace: true });
      return;
    }

    if (section === SICP_INDEX) {
      setSicpSectionLocalStorage(SICP_INDEX);
      return;
    }

    setLoading(true);

    fetch(baseUrl + section + extension)
      .then(response => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(myJson => {
        try {
          const newData = parseArr(myJson, refs); // Might throw error
          setData(newData);
          setSicpSectionLocalStorage(section); // Sets local storage if valid page
        } catch (error) {
          throw new ParseJsonError(error.message);
        }
      })
      .catch(error => {
        console.error(error);

        if (error.message === 'Not Found') {
          // page not found
          setData(getSicpError(SicpErrorType.PAGE_NOT_FOUND_ERROR));
        } else if (error instanceof ParseJsonError) {
          // error occurred while parsing JSON
          setData(getSicpError(SicpErrorType.PARSING_ERROR));
        } else {
          setData(getSicpError(SicpErrorType.UNEXPECTED_ERROR));
        }
        setLocalStorage(SICP_CACHE_KEY, SICP_INDEX); // Prevents caching invalid page
      })
      .finally(() => {
        setLoading(false);
      });
  }, [section, navigate, props]);

  // Scroll to correct position
  React.useEffect(() => {
    if (loading) {
      return;
    }

    const ref = refs.current[`#${hash}`];
    scrollRefIntoView(ref);
  }, [loading, hash]);

  // Close all active code snippet when new page is loaded
  React.useEffect(() => {
    setActive('0');
  }, [data]);

  const dispatch = useDispatch();
  const handleSnippetEditorOpen = (s: string) => {
    setActive(s);
    dispatch(WorkspaceActions.resetWorkspace('sicp'));
    dispatch(WorkspaceActions.toggleUsingSubst(false, 'sicp'));
  };
  const handleNavigation = (sect: string) => {
    // console.log(`Navigating to section: ${sect}`);
    setSection(sect);
  };

  // `section` is defined due to the navigate logic in the useEffect above
  const navigationButtons = (
    <div className="sicp-navigation-buttons" style={{ margin: '50px 0' }}>
      {getPrev(section!) && (
        <Button style={{ textWrap: 'nowrap' }} onClick={() => handleNavigation(getPrev(section!)!)}>
          Previous
        </Button>
      )}
      {getNext(section!) && (
        <Button style={{ textWrap: 'nowrap' }} onClick={() => handleNavigation(getNext(section!)!)}>
          Next
        </Button>
      )}
    </div>
  );

  const handleNodeClicked = React.useCallback(
    (node: TreeNodeInfo) => {
      // console.log(`Navigating to section: ${String(node.nodeData)}`);
      setSection(String(node.nodeData));
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [section]
  );

  const handleNodeClickedString = React.useCallback(
    (sect: string) => {
      const urlPart = sect.split('#');
      if (urlPart.length > 1) {
        console.log(`URL Part: ${urlPart[1]}`);
        setHash(urlPart[1]);
      }
      // console.log(`Navigating to section: ${urlPart[0]}`);
      setSection(urlPart[0]);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [section]
  );

  return (
    <div style={{ height: '100%' }} className={classNames('Sicp', Classes.TEXT_LARGE)}>
      <SicpNavigationBar
        handleNodeClickedString={handleNodeClickedString}
        handleNodeClicked={handleNodeClicked}
      />
      <SicpErrorBoundary>
        <CodeSnippetContext.Provider value={{ active: active, setActive: handleSnippetEditorOpen }}>
          {loading ? (
            <div className="sicp-content">{loadingComponent}</div>
          ) : section === 'index' ? (
            <SicpIndexPage handleNodeClick={handleNodeClicked} />
          ) : (
            <div className="sicp-content sicp-assessment">
              <Link to="#begin" ref={ref => (refs.current['#begin'] = ref)} />
              {data}
              {navigationButtons}
              <Link to="#end" ref={ref => (refs.current['#end'] = ref)} />
            </div>
          )}
        </CodeSnippetContext.Provider>
      </SicpErrorBoundary>
    </div>
  );
};

// react-router lazy loading
// https://reactrouter.com/en/main/route/lazy
export const Component = Sicp;
Component.displayName = 'Sicp';

export default Sicp;
