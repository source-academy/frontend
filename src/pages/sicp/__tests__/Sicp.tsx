import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import ReactRouter from 'react-router';
import { mockInitialStore } from 'src/commons/mocks/StoreMocks';

import Sicp from '../Sicp';
import SicpIndexPage from '../subcomponents/SicpIndexPage';

const routeComponentPropsMock = {
  history: {} as any,
  location: {} as any,
  match: {} as any
};

describe('Sicp renders', () => {
  test('correctly', () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ section: 'index' });

    const sicp = (
      <Provider store={mockInitialStore()}>
        <Sicp {...routeComponentPropsMock} />
      </Provider>
    );
    const tree = shallow(sicp);
    expect(tree.debug()).toMatchSnapshot();
  });

  test('index section correctly', () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ section: 'index' });
    window.HTMLElement.prototype.scrollIntoView = function () {};

    const sicp = (
      <Provider store={mockInitialStore()}>
        <Sicp {...routeComponentPropsMock} />
      </Provider>
    );
    const wrapper = mount(sicp);
    expect(wrapper.contains(<SicpIndexPage />)).toBeTruthy();
  });
});
