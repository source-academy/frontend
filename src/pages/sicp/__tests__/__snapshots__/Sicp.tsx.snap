// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sicp renders correctly 1`] = `
"<ContextProvider value={{...}}>
  <Sicp history={{...}} location={{...}} match={{...}} />
</ContextProvider>"
`;
